exporters/autoexport/registry.go

```diff
@@ -0,0 +1,149 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoexport // import "go.opentelemetry.io/contrib/exporters/autoexport"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+const (
+	otelExporterOTLPProtoEnvKey = "OTEL_EXPORTER_OTLP_PROTOCOL"
+)
+
+type traceSpanExporterFactoryFunc func(context.Context) (trace.SpanExporter, error)
+
+// registry maintains a map of exporter names to SpanExporter factories
+// func(context.Context) (trace.SpanExporter, error) that is safe for concurrent use by multiple
+// goroutines without additional locking or coordination.
+type registry struct {
+	mu    sync.Mutex
+	names map[string]traceSpanExporterFactoryFunc
+}
+
+func newRegistry() registry {
+	return registry{
+		names: map[string]traceSpanExporterFactoryFunc{
+			"":     buildOTLPExporter,
+			"otlp": buildOTLPExporter,
+		},
+	}
+}
+
+var (
+	// envRegistry is the package level registry of exporter registrations
+	// and their mapping to a SpanExporter factory func(context.Context) (trace.SpanExporter, error).
+	envRegistry = newRegistry()
+
+	// errUnknownExporter is returned when an unknown exporter name is used in
+	// the OTEL_*_EXPORTER environment variables.
+	errUnknownExporter = errors.New("unknown exporter")
+
+	// errInvalidOTLPProtocol is returned when an invalid protocol is used in
+	// the OTEL_EXPORTER_OTLP_PROTOCOL environment variable.
+	errInvalidOTLPProtocol = errors.New("invalid OTLP protocol - should be one of ['grpc', 'http/protobuf']")
+
+	// errDuplicateRegistration is returned when an duplicate registration is detected.
+	errDuplicateRegistration = errors.New("duplicate registration")
+)
+
+// load returns tries to find the SpanExporter factory with the key and
+// then execute the factory, returning the created SpanExporter.
+// errUnknownExporter is returned if the registration is missing and the error from
+// executing the factory if not nil.
+func (r *registry) load(ctx context.Context, key string) (trace.SpanExporter, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	factory, ok := r.names[key]
+	if !ok {
+		return nil, errUnknownExporter
+	}
+	return factory(ctx)
+}
+
+// store sets the factory for a key if is not already in the registry. errDuplicateRegistration
+// is returned if the registry already contains key.
+func (r *registry) store(key string, factory traceSpanExporterFactoryFunc) error {
```

Comment:
`traceSpanExporterFactoryFunc` should be exported if we expect users to provide their own implementations of it as arguments.

Reply:
How about renaming to `SpanExporterFactory`?

Reply:
Removed the custom func type -- it's not necessary and was only used in a few places.